2.2.0 compatibility guide (#4877)
JetBrains/kotlin-web-site: docs/topics/compatibility-guides/compatibility-guide-22.md

> If you want to add sources without replacing existing ones, use the [`source()`](https://kotlinlang.org/api/kotlin-gradle-plugin/kotlin-gradle-plugin-api/org.jetbrains.kotlin.gradle.tasks/-kotlin-compile-tool/source.html#) function.
>
> **Deprecation cycle**:
>
> - 2.2.0: enable new behavior

### Deprecate KotlinCompilationOutput#resourcesDirProvider

> **Issue**: [KT-70620](https://youtrack.jetbrains.com/issue/KT-70620)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: The `KotlinCompilationOutput#resourcesDirProvider` field is deprecated.
> Use `KotlinSourceSet.resources` in your Gradle build script instead to add additional resource directories.
>
> **Deprecation cycle**:
>
> - 2.1.0: `KotlinCompilationOutput#resourcesDirProvider` is deprecated with a warning
> - 2.2.0: raise the warning to an error

### Change the type of annotationProcessorOptionProviders to use CommandLineArgumentProvider

> **Issue**: [KT-58009](https://youtrack.jetbrains.com/issue/KT-58009)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>

> > **Short summary**: The type of [`annotationProcessorOptionProviders`](https://kotlinlang.org/api/kotlin-gradle-plugin/kotlin-gradle-plugin-api/org.jetbrains.kotlin.gradle.tasks/-base-kapt/annotation-processor-option-providers.html#) in `BaseKapt` is changed from `MutableList<Any>`

Reviewer: Not true - `BaseKapt.annotationProcessorOptionProviders` still has `MutableList<Any>` type and this property is deprecated.

Users should migrate to `BaseKapt.annotationProcessorOptionsProviders` (note: option**s**) which has `ListProperty<CommandLineArgumentProvider>` type.

> to `MutableList<CommandLineArgumentProvider>`. 
> This clearly defines the expected element type and prevents runtime failures caused by adding incorrect elements, such as nested lists.
> If your code currently adds a list as a single element, use the `addAll()` function instead of the `add()` function.
>
> **Deprecation cycle**:
>
> - 2.2.0: enforce new type in the API

### Deprecate the `kotlin-android-extensions` plugin

> **Issue**: [KT-72341](https://youtrack.jetbrains.com/issue/KT-72341/)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: The `kotlin-android-extensions` plugin is deprecated. Use a separate plugin, `kotlin-parcelize`,
> for the `Parcelable` implementation generator and the Android Jetpack's [view bindings](https://developer.android.com/topic/libraries/view-binding)
> for synthetic views instead.
>
> **Deprecation cycle**:
>
> - 1.4.20: the plugin is deprecated
> - 2.1.20: a configuration error is introduced, and no plugin code is executed

> > - 2.2.20: the plugin is removed

Reviewer: ```suggestion
> - 2.2.0: the plugin is removed
```

> `BaseKapt.annotationProcessorOptionsProviders`, which accepts a `ListProperty<CommandLineArgumentProvider>` instead of a `MutableList<Any>`.
> This clearly defines the expected element type and prevents runtime failures caused by adding incorrect elements, such as nested lists.
> If your current code adds a list as a single element, replace the `add()` function with the `addAll()` function.
>
> **Deprecation cycle**:
>
> - 2.2.0: enforce the new type in the API

### Deprecate `kotlin-android-extensions` plugin

> **Issue**: [KT-72341](https://youtrack.jetbrains.com/issue/KT-72341/)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: The `kotlin-android-extensions` plugin is deprecated. Use a separate plugin, [`kotlin-parcelize`](https://plugins.gradle.org/plugin/org.jetbrains.kotlin.plugin.parcelize),
> for the `Parcelable` implementation generator and the Android Jetpack's [view bindings](https://developer.android.com/topic/libraries/view-binding)
> for synthetic views instead.
>
> **Deprecation cycle**:
>
> - 1.4.20: the plugin is deprecated
> - 2.1.20: a configuration error is introduced, and no plugin code is executed

> > - 2.2.0: the plugin is removed

Reviewer 2: ```suggestion
> - 2.2.0: the plugin code is removed
```

> This affects cases where Java classes extend Kotlin ones and when working with mapped types like `java.util.LinkedHashSet`.
> To migrate, replace property access with direct calls to the corresponding getter functions.
>
> **Deprecation cycle**:
>
> - 2.0.0: report a warning for accessing synthetic properties created from Kotlin getters
> - 2.2.0: raise the warning to an error

### Change default method generation for interface functions on JVM

> **Issue**: [KTLC-269](https://youtrack.jetbrains.com/issue/KTLC-269)
>
> **Component**: Core language
>
> **Incompatible change type**: binary
>
> **Short summary**: Functions declared in interfaces are now compiled to JVM default methods unless configured otherwise.
> This may result in compilation errors in Java code when unrelated supertypes define conflicting implementations.
> The behavior is controlled by the stable `-jvm-default` compiler option, which replaces the now deprecated `-Xjvm-default` option.
> To restore the previous behavior, where default implementations are generated only in `DefaultImpls` classes and subclasses, use `-jvm-default=disable`.
>
> **Deprecation cycle**:
>

> > - 2.2.0: `-jvm-default` compiler option is set to `enable` by default

Author: hmm 🤔 - it was introduced to the[ Gradle compiler options doc ](https://kotlinlang.org/docs/gradle-compiler-options.html)and not to the generic compiler-reference. Should it also be included there? 🤔

> Use `KotlinSourceSet.resources` in your Gradle build script instead to add additional resource directories.
>
> **Deprecation cycle**:
>
> - 2.1.0: `KotlinCompilationOutput#resourcesDirProvider` is deprecated with a warning
> - 2.2.0: raise the warning to an error

### Deprecate `BaseKapt.annotationProcessorOptionProviders` property

> **Issue**: [KT-58009](https://youtrack.jetbrains.com/issue/KT-58009)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: The [`BaseKapt.annotationProcessorOptionProviders`](https://kotlinlang.org/api/kotlin-gradle-plugin/kotlin-gradle-plugin-api/org.jetbrains.kotlin.gradle.tasks/-base-kapt/annotation-processor-option-providers.html#) property is deprecated in favor of 
> `BaseKapt.annotationProcessorOptionsProviders`, which accepts a `ListProperty<CommandLineArgumentProvider>` instead of a `MutableList<Any>`.
> This clearly defines the expected element type and prevents runtime failures caused by adding incorrect elements, such as nested lists.
> If your current code adds a list as a single element, replace the `add()` function with the `addAll()` function.
>
> **Deprecation cycle**:
>

> > - 2.2.0: enforce new type in the API

Reviewer 3: ```suggestion
> - 2.2.0: enforce the new type in the API
```

### Deprecate `kotlin-android-extensions` plugin

> **Issue**: [KT-72341](https://youtrack.jetbrains.com/issue/KT-72341/)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>

> > **Short summary**: The `kotlin-android-extensions` plugin is deprecated. Use a separate plugin, `kotlin-parcelize`,

Reviewer 3: Add link to the parcelize plugin? https://plugins.gradle.org/plugin/org.jetbrains.kotlin.plugin.parcelize

> for the `Parcelable` implementation generator and the Android Jetpack's [view bindings](https://developer.android.com/topic/libraries/view-binding)
> for synthetic views instead.
>
> **Deprecation cycle**:
>
> - 1.4.20: the plugin is deprecated
> - 2.1.20: a configuration error is introduced, and no plugin code is executed

> > - 2.2.0: the plugin is removed

Reviewer 3: Does this mean support for the plugin is removed in Kotlin?

### Deprecate `kotlinOptions` DSL

> **Issue**: [KT-54110](https://youtrack.jetbrains.com/issue/KT-54110)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: The ability to configure compiler options through the `kotlinOptions` DSL and the related
> `KotlinCompile<KotlinOptions>` task interface is deprecated in favor of the new `compilerOptions` DSL.
> As part of this deprecation, all properties in the `kotlinOptions` interface are now also individually marked as deprecated.

> > To migrate, use the `compilerOptions` DSL to configure compiler options.

Reviewer 3: ```suggestion
> To migrate, use the `compilerOptions` DSL to configure compiler options. For guidance on the migration, see [Migrate from `kotlinOptions {}` to `compilerOptions {}`](gradle-compiler-options.md#migrate-from-kotlinoptions-to-compileroptions).
```

>
> * `useDisambiguationClassifierAsSourceSetNamePrefix`
> * `overrideDisambiguationClassifierOnIdeImport`
>
> **Deprecation cycle**:
>
> - 2.0.0: report a warning when the Gradle properties are used
> - 2.1.0: raise this warning to an error
> - 2.2.0: remove Gradle properties

### Deprecate commonization parameters

> **Issue**: [KT-75161](https://youtrack.jetbrains.com/issue/KT-75161)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: The parameters for experimental commonization modes are deprecated in the Kotlin Gradle plugin.
> These parameters may produce invalid compilation artifacts that are then cached. To delete affected artifacts:
>
> 1. Remove the following options from your `gradle.properties` file:
>
>    ```none
>    kotlin.mpp.enableOptimisticNumberCommonization
>    kotlin.mpp.enablePlatformIntegerCommonization
>    ```
>
> 2. Clear the commonization cache in the `~/.konan/*/klib/commonized` directory or run the following command: 
>
>    ```bash
>    ./gradlew cleanNativeDistributionCommonization
>    ```
>
> **Deprecation cycle**:
>

> > - 2.2.0: commonization parameters are deprecated with an error

Reviewer 3: ```suggestion
> - 2.2.0: derprecate commonization parameters with an error
```

> - 2.2.20: remove commonization parameters

### Deprecate support for legacy metadata compilation

> **Issue**: [KT-61817](https://youtrack.jetbrains.com/issue/KT-61817)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: Options that were used to set up hierarchical structure and create intermediate source sets
> between the common and intermediate source sets became obsolete.
> The following compiler options are removed:
> 
> * `isCompatibilityMetadataVariantEnabled`
> * `withGranularMetadata`
> * `isKotlinGranularMetadataEnabled`
>
> **Deprecation cycle**:

> 

Reviewer 3: ```suggestion
>
```

> - 2.2.0: remove compiler options from the Kotlin Gradle plugin

### Deprecate `KotlinCompilation.source`

> **Issue**: [KT-64991](https://youtrack.jetbrains.com/issue/KT-64991)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>

> > **Short summary**: The access to `KotlinCompilation.source` that allowed adding Kotlin source sets directly to the Kotlin

Reviewer 3: @danil-pavlov do you mean that the `source()` function is deprecated? It's  not clear what "the access" means for me 🤔

Reviewer 2: This is actually a deprecation cycle update of [this change](https://www.jetbrains.com/help/kotlin-multiplatform-dev/multiplatform-compatibility-guide.html#deprecated-api-for-adding-kotlin-source-sets-directly-to-the-kotlin-compilation), so I tried to be consistent with that text
Let's call it `KotlinCompilation.source` API? to be more general

> compilation has been deprecated.
>
> **Deprecation cycle**:
>

> > - 1.9.0: report a warning when `KotlinComplation.source` is used

Reviewer 3: ```suggestion
> - 1.9.0: report a warning when `KotlinCompilation.source` is used
```

> - 1.9.20: raise this warning to an error

> > - 2.2.0: remove `KotlinComplation.source` from the Kotlin Gradle plugin; attempts to use it lead to "unresolved reference"

Reviewer 3: ```suggestion
> - 2.2.0: remove `KotlinCompilation.source` from the Kotlin Gradle plugin; attempts to use it lead to "unresolved reference"
```

> when multiple flavors and build types are used. Use the `publishLibraryVariants()` function, which specifies build variants,
> instead.
>
> **Deprecation cycle**:
>
> - 2.2.0: `publishAllLibraryVariants()` is deprecated

### Deprecate `android` target

> **Issue**: [KT-71608](https://youtrack.jetbrains.com/issue/KT-71608)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: The `android` target name is deprecated in the current Kotlin DSL. Use `androidTarget` instead.
>
> **Deprecation cycle**:
>
> - 1.9.0: introduce a deprecation warning when the `android` name is used in Kotlin Multiplatform projects 
> - 2.1.0: raise this warning to an error
> - 2.2.0: remove the `android` target DSL from the Kotlin Multiplatform Gradle plugin

### Deprecate `konanVersion` in `CInteropProcess`

> **Issue**: [KT-71069](https://youtrack.jetbrains.com/issue/KT-71069)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: The `konanVersion` property in the `CInteropProcess` task is deprecated.
> Use `CInteropProcess.kotlinNativeVersion` instead.
>
> **Deprecation cycle**:
>

> > - 2.1.0: report a warning when `konanVersion` is used

Reviewer 3: ```suggestion
> - 2.1.0: report a warning when the `konanVersion` property is used
```

> - 2.2.0: raise this warning to an error

> > - 2.3.0: remove `konanVersion` property from the Kotlin Gradle plugin

Reviewer 3: ```suggestion
> - 2.3.0: remove the `konanVersion` property from the Kotlin Gradle plugin
```

### Deprecate `destinationDir` in `CInteropProcess`

> **Issue**: [KT-71068](https://youtrack.jetbrains.com/issue/KT-71068)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: The `destinationDir` property in the `CInteropProcess` task is deprecated.

> > Use `CInteropProcess.destinationDirectory.set()` instead.

Reviewer 3: ```suggestion
> Use the `CInteropProcess.destinationDirectory.set()` function instead.
```

>
> **Deprecation cycle**:
>

> > - 2.1.0: report a warning when `destinationDir` is used

Reviewer 3: ```suggestion
> - 2.1.0: report a warning when the `destinationDir` property is used
```

> - 2.2.0: raise this warning to an error

> > - 2.3.0: remove `destinationDir` from the Kotlin Gradle plugin

Reviewer 3: ```suggestion
> - 2.3.0: remove the `destinationDir` property from the Kotlin Gradle plugin
```

### Deprecate `kotlinArtifacts` API

> **Issue**: [KT-74953](https://youtrack.jetbrains.com/issue/KT-74953)
>
> **Component**: Gradle
>
> **Incompatible change type**: source
>
> **Short summary**: The experimental `kotlinArtifacts` API is deprecated. Use the current DSL available in the Kotlin Gradle
> plugin to [build final native binaries](https://www.jetbrains.com/help/kotlin-multiplatform-dev/multiplatform-build-native-binaries.html).

> > If it's not sufficient for migration, leave a comment in [this YT issue](https://youtrack.jetbrains.com/issue/KT-74953).

Reviewer 3: @danil-pavlov by "it" do you mean the current DSL?

Reviewer 2: yes, correct

>
> **Deprecation cycle**:
>

> > - 2.2.0: report a warning when `kotlinArtifacts` is used

Reviewer 3: ```suggestion
> - 2.2.0: report a warning when the `kotlinArtifacts` API is used
```
